### ISSUE TITLE
Add ability to force using system-installed dependency instead of Conan package

### ISSUE THREAD

**Issue by GamePad64** [conan-io/conan#1330]
Hello, I am writing a Qt-based library and want to create a Conan package.
But, I want to give the end-user a choice:
- install Qt with Conan, as a usual Conan dependency

OR

- make Conan assume, that Qt is already installed by the platform-dependent package manager (apt, rpm, homebrew, ... whatever), and rebuild the package from source (to mitigate the possible version incompatibility).

But, I want this mechanism to be more generic and standardized, so the end-user could mark any dependency to be "system-installed" (using Conan profile, for example) and all packages, dependent on this will be rebuilt, assuming, that this dependency is present on the system.

**Comment by memsharded** [comment]
Hi Alexander,

I think there are couple of different approaches for this:

- Have a different Qt recipe, that acts only as a wrapper for the system installed one. It might take advantage of the ``system_requirements()`` method and the ``SystemPackageTool`` helper: http://conanio.readthedocs.io/en/latest/reference/tools.html?highlight=SystemPackageTool#tools-osinfo-and-tools-systempackagetool. Put that package under the **system** channel. Then you would have to different versions for your Qt/version library, one would be the full conan one, able to build from source code and manage the binaries within conan, and the other would be just the wrapper. Packages can then use conditional dependencies to depend on one or the other, or use dependency overriding from downstream, to force the usage of the one you want.
- Make your Qt recipe, configurable with an option, something like ``system=[True, False]``, and use that option inside the recipe to act accordingly. ``build()`` and ``package()`` methods would do no thing for ``system=True``.

I think both could be valid solutions, depending on the personal preferences and particular team workflows, would choose one or the other. What do you think? Could be a solution?

**Comment by hoxnox** [comment]
I tried to move through the second approach (see my recipes, for example https://github.com/hoxnox/conan-libressl), but I had to use my own extended ConanFile (see https://github.com/hoxnox/conan-nxtools) to make `system` and `root` options available in all my packages.

**Comment by GamePad64** [comment]
@memsharded First approach is using native override mechanism, it appears difficult to use on an everyday basis: there are too many libs out there inside the system, and writing a separate (dummy) conanfile.py for each of them I want to override will lead to a great clutter.

@hoxnox You are moving in a right direction, and all your packages can be overridden by the system-specific package. I propose adding this functionality directly into Conan, so every package could be overridden.

**Note:** maybe, restrict this option to _package consumers_, so Conan packages would depend only on other Conan packages (not on the system ones). Then, every package could still be installed on a clean system using a full dependency graph without making any assumptions about system-installed libs.

**Comment by lasote** [comment]
To summarize, are you suggesting something like an always available option `system` defaulted to `False`and a `package_info_system()` method that will just declare the system's library names and any other flag needed or I'm missing something?

**Comment by hoxnox** [comment]
There are situations in wich there may be several system packages with the same name (very often on WIndows). So there should be ability to force `root directory` helped to find the lib. In my code it looks like that:

```python
    def package_info(self):
        self.do_package_info()
        if self.options.system:
            self.cpp_info.includedirs = []
            self.cpp_info.libdirs = []
            if len(str(self.options.root)) != 0:
                self.cpp_info.includedirs.append(str(self.options.root) + "/include")
                self.cpp_info.libdirs.append(str(self.options.root) + "/lib")
```

package_info() and package_info_system() should not conflict. Who is responsible for writing package_info_system()? Package creator? Or package user? What if package creator know nothing about system option...

**Comment by lasote** [comment]
But if you think that the package user has to take the decision about which system package link and the root...(I see it reasonable) It's almost the same effort than create a dummy recipe and override it, right?

**Comment by GamePad64** [comment]
Personally, I am thinking only about an always available option "system" or "skip", that basically skips the build(), source(), package() and import() steps from the package., without any package_info_system() methods.

**Comment by GamePad64** [comment]
@lasote basically, yes, but without writing dummy recipes by hand.

**Comment by hoxnox** [comment]
For complex situations (root change) dummy package should be created, for simple cases - option `system` would be great.

**Comment by ZaMaZaN4iK** [comment]
@lasote @memsharded what is the status of the issue?

**Comment by memsharded** [comment]
For me, it is still confusing this feature request.

Lets start with the UI. Lets say we have some package "LibA", with a conanfile.py and a conan package we can install, and such LibA is also available by other means at the system level.

Lets say that we implement a command-line ``skip`` option, it would be something like this?

```bash
$ conan install .. --skip=LibA
```
And then, even if I have listed ``LibA/version@user/channel`` in my conanfile, or it is in my transitive dependencies, it won't get installed at all, because it would be at the system level? Or I get installed the recipe, but it does nothing, but the ``package_info()``? But ``package_info()`` will be most likely incompatible/wrong for the system packages.

So there are a few questions:
- If I created a LibB conanfile.py package that depends on LibA, that of course uses the conan information (e.g. with the cmake generator and the ``conanbuildinfo.cmake``) to link with it, it will fail if I provide the ``skip`` option, right? 
- I could add some manual stuff to LibB build scripts, but that doesn't scale at all if I have some other packages LibC, LibD, etc that depends on LibA
- What happens if LibA has other transitive dependencies. I don't think this is possible at all, so the suggested approach is only for packages with no transitive dependencies? That would quickly break, even for system packages. Gathering the transitive dependencies information for all system package managers is totally unfeasible, it is not realistic to implement that in conan.

Please @GamePad64 @hoxnox @lasote feedback.

**Comment by hoxnox** [comment]
> If I created a LibB conanfile.py package that depends on LibA, that of course uses the conan information (e.g. with the cmake generator and the conanbuildinfo.cmake) to link with it, it will fail if I provide the skip option, right?

Why? libA package should find libs and headers in the system and provide correct package_info, so libB could use the provided information for link with.

> I could add some manual stuff to LibB build scripts, but that doesn't scale at all if I have some other packages LibC, LibD, etc that depends on LibA

Some manual stuff should be added in libA package. It will not affect libB, libC and libD.


As for me - the new argument doesn't needed. Just wide ConanPackage with default `system` and `root` options, for example like I [do in my packages](https://github.com/hoxnox/conan-nxtools/blob/master/nx_conan_file.py)  and describe them in the documentation to force package creators use it. This helps a lot with building packages for various OS (windows, debian, gentoo).

**Comment by memsharded** [comment]
@hoxnox Your approach is very interesting indeed, as it is related to the "conan project" information. Please have a look: https://github.com/conan-io/conan/issues/1377

The idea was to be able to define in a file a set of mappings, from PkgName->Root folder. This allows  not only for system dependencies, but also for user dependencies in user space. That ``root`` approach in "conan-project" is already implemented as a PoC and does not require to code or modify at all the package recipes.

**Comment by clssn** [comment]
Hey @memsharded , please allow me to put my use-case into the mix, hoping to revive this discussion.

I've got a library A requiring boost v1.69.0 (conveniently getting it from Conan Center).
A plugin library B to be loaded at runtime by A has a dependency to ROS (Robot Operating System). ROS is quite popular, but I *can't find ROS libraries in any Conan repository*. So I have to use the one installed on Ubuntu (still 18.04, so ROS melodic).
Using a `find_package(ros_cpp REQUIRED)` this is not a problem. Problem is though, that ROS melodic depends on the older boost 1.65.1 whose include directory is `/usr/include`. Building package B CMake gets the boost 1.69.0 include directory first in the include list, effectively hiding the system's boost 1.65.1. Unfortunately I need to use 1.65.1 here because when I compile with the 1.69.0 headers, I get a segfault at runtime. 

**tl;dr**: In package B I'd like to override the transitional boost 1.69.0 requirement from package A with the system installed one for use in package B. But I don't want to maintain a whole lot of wrapper packages for these kinds of things. 

Bottom line: In an ideal world, I would get ROS via Conan and have Conan handle dependencies one or the other way, but as long as there are libs that we need to use from the system, I'd like to have a means to express that with Conan.

**Comment by solosuper** [comment]
Hi @clssn, what is your current approach for this? Your post describes my exact problem too, being forced to work on Ubuntu 18.04 and trying to mix ROS melodic and conan.

So far it has worked ok with `boost/1.65.1@conan/stable", override=True`, but with the sunset of bintray that package can no longer be built from source (thus I have a local patch of the conanfile.py that downloads the sources from sourceforge instead).

I startet to think about writing a wrapper for the apt-installed boost, which brought me here. Even after reading this thread, it's not entirely clear for me how one should write such a wrapper, so if anyone could provide a minimal example (or a link to one), I will be grateful.

**Comment by skycaptain** [comment]
Similar use-case here. We're using Yocto and the Yocto SDK to cross-compile for an embedded device but also Conan to build for Linux, Windows and MacOS. Furthermore, we're using CMake and `find_package`. Currently, we implemented this by using Conan only for Linux, Windows and MacOS; and CMake directly for the Yocto SDK, however with the latter we have to make sure every dependency is delivered also with the right version with the SDK, which is not always the case. Consider the case where you have MyAppA -> MyLibB -> 3rdPartyLibC. When a breaking change is made to MyLibB, one would first have to commit the changes to MyLibB, integrate the changes in Yocto (which would not work in the first place, as at the same time they break MyAppA, hence we would have a broken pipeline), build a new Yocto SDK, use the new SDK to fix and test MyAppA and then integrate that into Yocto. One solution to overcome this is not to deliver our own libraries (i.e MyLibB) with the SDK at all, but only third-party (3rdPartyLibC), which rarely change.

Hence, we basically maintain three systems:
- Building with Yocto, which is great for integration but not for development.
- Building with the Yocto SDK and CMake directly, where we have to do the dependency management ourselves.
- Building with Conan only for Linux and Windows, where Conan handles the dependencies.

We want to get rid of the second one. We think of having a mechanism to tell Conan to assume a package is already available, i.e. skip building, as specific packages are already shipped with the Yocto SDK. Missing packages should be built on-demand.

[This article](https://docs.conan.io/en/latest/integrations/cross_platform/yocto.html) from conan-docs basically says to use the SDK to compile every package required with Conan (incl. 3rdPartyLibC), push everything to Artifactory and consume the pre-built artifacts in the Yocto build. However, we want to use it the other way around: Build everything third-party plus the toolchain into the SDK, and build with Conan what's missing.

**Comment by kambala-decapitator** [comment]
I want to use some Apple system libraries (e.g. zlib, iconv) instead of using external versions. I created wrapper like this:

```py
from conan import ConanFile

class IconvConan(ConanFile):
    name = "libiconv"
    version = "1.17"

    def package_info(self):
        self.cpp_info.includedirs = []
        self.cpp_info.system_libs = ["iconv"]
```

And in my project's `conanfile.py` I use it like this:

```py
def requirements(self):
    if tools.is_apple_os(self.settings.os):
        systemLibsOverrides = [
            "libiconv/1.17",
            "zlib/1.2.12",
        ]
        for lib in systemLibsOverrides:
            self.requires(f"{lib}@kambala/apple", override=True)
```

This works great with boost, for example:

```
> otool -L /Users/kambala/.conan/data/boost/1.79.0/_/_/package/cc3594ca98fe965985efe959cf277972652a55fa/lib/libboost_locale.dylib 
.../libboost_locale.dylib:
	@rpath/libboost_locale.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.23.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
```

But with libpng it doesn't -- the system library isn't propagated:

```
[7/36] Linking C executable bin/png-fix-itxt
FAILED: bin/png-fix-itxt 
: && /Applications/Xcode13.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -m64 -O3 -DNDEBUG  -arch x86_64 -isysroot /Applications/Xcode13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -mmacosx-version-min=10.13 -Wl,-search_paths_first -Wl,-headerpad_max_install_names  source_subfolder/CMakeFiles/png-fix-itxt.dir/contrib/tools/png-fix-itxt.c.o -o bin/png-fix-itxt   && :
Undefined symbols for architecture x86_64:
  "_crc32", referenced from:
      _main in png-fix-itxt.c.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

as you can see, `-lz` flag isn't passed.

I've inspected libpng's CMakeLists and it seems that the reason is that it links to `${ZLIB_LIBRARIES}` directly instead of the `zlib::zlib` target. Is it possible to overcome that without applying patches to libpng, i.e. adjust my wrapper somehow? Tried adding

    self.cpp_info.set_property("cmake_find_mode", "none")

as [CMakeDeps](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html#properties) advises but it didn't help.

Posting it here as I spent a considerable amount of time trying to figure out how to create system library wrapper, it might help someone else.

**EDIT**: found it!

```py
self.cpp_info.sharedlinkflags = ["-lz"]
self.cpp_info.exelinkflags = ["-lz"]
```

**Comment by memsharded** [comment]
Hi @kambala-decapitator 

One of the important issues here is trying to use recipes from ConanCenter, that assumes that their transitive dependencies are other Conan packages, together with your own variants that implement the logic to depend on the system ones instead.

As a hint for other readers, it is necessary to be consistent, irrespective of the approach:
- If you make conditional logic to depend on zlib as a Conan package, that conditional logic has to be in all other Conan recipes that consume zlib, it cannot be in your own ones, but still uses packages from ConanCenter as-is, because they will depend on zlib Conan package
- If you use a Conan package wrapper around zlib, it is easier for the consumers, that don't need to be modified, but you need to correctly implement the conditional logic in the zlib Conan recipe ``conanfile.py`` instead, building the library in some cases, and using the system one, with declarations as the ``self.cpp_info.sharedlinkflags = ["-lz"]`` in others.

Note: I think it would be better to use the ``self.cpp_info.system_libs = ["z"]`` instead

**Comment by kambala-decapitator** [comment]
@memsharded thank you for the explanation!

My use case is using Conan to install dependencies for an application. I think it's fine to use "local" overrides here, as there're no consumers of the "application package". But I fully agree that this is not the way to go when working on a package that should be consumed by the others: ideally, the system/package lib selection should be inside the official conan recipe, which this issue is exactly about if I understand correctly.

> Note: I think it would be better to use the `self.cpp_info.system_libs = ["z"]` instead

yes, that's what I used originally, as visible in my iconv package example. But this approach didn't work with libpng, that's why I also had to inject linker flags explicitly. Now my custom recipes look like this:

```py
from conan import ConanFile

class ZlibConan(ConanFile):
    name = "zlib"
    version = "1.2.12"

    def package_info(self):
        self.cpp_info.includedirs = []

        libName = "z"
        self.cpp_info.system_libs = [libName]
        self.cpp_info.sharedlinkflags = [f"-l{libName}"]
        self.cpp_info.exelinkflags = [f"-l{libName}"]
```

P.S. But even this didn't work with PCRE2's `pcre2grep`. Short check of [the recipe](https://github.com/conan-io/conan-center-index/tree/f3e554f4e35d4c406536f8bd291cf0b90e20fe71/recipes/pcre2/all/conanfile.py#L173) showed that there's some workaround applied already, but I didn't investigate deeply why it didn't work and ended up by simply disabling pcre2grep build with `"pcre2/*:build_pcre2grep": False`.

**Comment by memsharded** [comment]
Conan 2.0 provides the ``system_tool`` approach: https://docs.conan.io/2/reference/config_files/profiles.html#system-tools, that can be used to force system installed versions for things like cmake, meson, or compiler. 

But basically only things that are very simple to use and install, like just being in the PATH works fine.
For any other case, like regular libraries, we have been re-checking, and it seems extremely unlikely that things will work transparently, because in most cases there is information captured by the recipe ``package_info()`` that is necessary to consume it portably. We might consider something similar to the ``system_tool`` for dependencies, but we are very concerned that this would probably only bring more pain and problems and will be more difficult to maintain than writing wrapper recipes.

**Comment by kambala-decapitator** [comment]
let me share recipes for some of system libs (not sure about their v2 compatibility), maybe someone would find them useful: https://github.com/kambala-decapitator/conan-system-libs

**Comment by memsharded** [comment]
We implemented this feature for ``tool_requires``, which can be defined in the profile as ``[system_tools]`` in Conan2.0
Have been considering to extend this to regular requires, but so far there are more unknowns and concerns than expected advantages, but lets see.

**Comment by memsharded** [comment]
This has been closed by https://github.com/conan-io/conan/pull/14871, for next 2.0.15.

Still there is another potential part for this in https://github.com/conan-io/conan/pull/15136, which includes not only using system-installed deps, but replacing requirements by other Conan recipes that would wrap system deps, please track that one.

Looking forward your feedback.
